### DISCUSSION TITLE
Page Object Functionality with MultiRemotes

### DISCUSSION THREAD

**Discussion by theo-rogers** [webdriverio/webdriverio#7623] [Q&A] [answered]
Is anyone aware of any documentation or an example project that shows how the PageObject model could work with MultiRemotes?
Hopefully something with Typescript, but I'm grateful for whatever I can find. I've been scouring for more information for about a week now.

**Comment by christian-bromann** [comment]
I don't think building POs with multiremote is anything different from running them in normal mode. Does what we described in [our docs](https://webdriver.io/docs/pageobjects) not work for you?

**Reply by gurman-eez** [reply]
> I don't think building POs with multiremote is anything different from running them in normal mode. Does what we described in [our docs](https://webdriver.io/docs/pageobjects) not work for you?

I have the same situation as @alexrogersdesign.

I think that he is talking about something like this:
```
// this will work only for browserA
await LoginPage.browserA.username.setValue('Max Mustermann')
//or
await BrowserAClass.LoginPage.username.setValue('Max Mustermann')

// this will work only for browserB
await LoginPage.browserB.username.setValue('Max Mustermann')
//or
await BrowserBClass.LoginPage.username.setValue('Max Mustermann')
```

**Reply by theo-rogers** [reply]
Yup pretty much. I wrote the workaround I ultimately came up with below. It's not an optimal solution, but hopefully it helps.

**Comment by theo-rogers** [comment, marked as answer]
@christian-bromann From what I can tell you cant distinguish between individual browsers when using POs (without doing anything hacky). What's written in the docs would work for controlling all browsers, but individual browser actions would not be able to use the functionality of the PO. 
As a workaround, I was able to add a browser instance variable to the PO class and pass the specific browser I wanted to control via a constructor. A PO could then be instantiated to represent the individual browser. This ends up leading to some code complexity issues because you now need to instantiate an object for each individual browser in addition to one to control all of the browsers. 
So a mocha test would look like this: 

````
describe('Join Meeting Multi-Remote', async () => {
  let multiPreMeetingPage;
  let multiPreMeetingPageA;
  let multiPreMeetingPageB;

  before('setup', async () => {
    const browser = await getMultiBrowser();
    multiPreMeetingPageA= preMeetingPageCreator(browser.a);
    multiPreMeetingPageB= preMeetingPageCreator(browser.b);
    multiPreMeetingPage = preMeetingPageCreator(browser);
    await multiPreMeetingPage.open();
    await multiPreMeetingPageA.login(userAEmail, 'test');
    await multiPreMeetingPageB.login(userBEmail, 'test');
  });
  after('logout', async () => multiPreMeetingPage.logout());

  it('a user should be notified when a another user joins', async () => {
    await multiPreMeetingPageA.joinFirstMeeting();
    await b.pause(1000);
    await multiPreMeetingPageB.joinFirstMeeting();
    const multiMeetingPageA = meetingPageCreator(a);
    await a.pause(4000);
    const elem = await multiMeetingPageA.userJoinedNotification;
    await elem.waitForDisplayed();
    const noteText = await elem.getText();
    expect(noteText).toContain(userBName);
  });
});
````

The PO now looks like this 

````
export default class Page {
 protected browser: any

 constructor(newBrowser?: ParamBrowser) {
   this.browser = newBrowser?? browser;
 }
 async open(path?: string) {
   return this.browser.url(path?? this.url);

````
Specific PO 
LoginPageCreator doesn't need to exist, it was originally implemented to not break the traditional PO model (and my single tests) and allow the PO to be instantiated in the export. I had to abandon this. 
````
class LoginPage extends Page {

  constructor(browser?:ParamBrowser) {
    super(browser);
  }
  
  ~~~~~~~~~~~~~~ other code ~~~~~~~~~~~~~~
  
  export function loginPageCreator(newBrowser?: ParamBrowser) {
  return new LoginPage(newBrowser);
}

export default LoginPage;

````



I also got it to work with Typescript (without switching to standalone), but with a pretty nasty typecast. 

````
export async function getMultiBrowser() {
  return browser as unknown as WebdriverIO.MultiRemoteBrowser;
}
````

One of the downsides of the approach was that I was not able to create a type compatible between both a standard browser and multiremote browser objects. I had to type the browser as any. Someone more skilled in Typescript or more familiar with the codebase could likely create a type with just the compatible constituents types needed for this approach. 

````
browser: WebdriverIO.MultiRemoteBrowser | WebdriverIO.Browser
````
Would not work etc.

**Comment by fspinillo** [comment]
@alexrogersdesign `As a workaround, I was able to add a browser instance variable to the PO class and pass the specific browser I wanted to control via a constructor. ` this is what I was going to suggest you try. This is how I was able to get around in my trialing of the feature.
